File: 1448. Count Good Nodes in Binary Tree/main.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(root: TreeNode, maxVal: int) -> int:
            good = 0

            if not root:
                return good

            if root.val >= maxVal:
                good = 1
                maxVal = root.val

            return good + dfs(root.left, maxVal) + dfs(root.right, maxVal)

        return dfs(root, -float("inf"))

File: 1448. Count Good Nodes in Binary Tree/main.cpp
#include <limits>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode()
        : val(0)
        , left(nullptr)
        , right(nullptr) {}
    TreeNode(int x)
        : val(x)
        , left(nullptr)
        , right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x)
        , left(left)
        , right(right) {}
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int dfs(TreeNode *root, int maxVal) {
        int good = 0;

        if (!root) {
            return good;
        }

        if (root->val >= maxVal) {
            good++;
            maxVal = root->val;
        }

        return good + dfs(root->left, maxVal) + dfs(root->right, maxVal);
    }

public:
    int goodNodes(TreeNode *root) {
        return dfs(root, numeric_limits<int>::min());
    }
};